server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java

```diff
@@ -59,6 +60,7 @@ public class QueryRewriteContext {
     protected boolean allowUnmappedFields;
     protected boolean mapUnmappedFieldAsString;
     protected Predicate<String> allowedFields;
+    private final Map<String, IndexMetadata> indexMetadataMap;
```

Comment:
I think we can introduce a new coordinator rewrite context. The existing one is for the special case of can match phase that rewrite per index in the coordinator so we can come up with a new name and/or change the name of the existing one. This will make it more clear and would not require to have these accessors that can be null or empty depending where the execution is done.

Reply:
Agreed, but I think this refactoring is best addressed in a follow-up PR